src/librustc/mir/interpret/mod.rs

```diff
@@ -575,6 +575,144 @@ impl<Tag, Extra: Default> Allocation<Tag, Extra> {
+            offset,
+            offset + size,
+        ).or_else(|idx| err!(ReadUndefBytes(idx)))
+    }
+
+    pub fn check_relocations(
+        &self,
+        hdl: impl HasDataLayout,
+        offset: Size,
+        size: Size,
+    ) -> EvalResult<'tcx> {
+        if self.relocations(hdl, offset, size)?.len() != 0 {
+            err!(ReadPointerAsBytes)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn relocations(
+        &self,
+        hdl: impl HasDataLayout,
+        offset: Size,
+        size: Size,
+    ) -> EvalResult<'tcx, &[(Size, (Tag, AllocId))]> {
+        // We have to go back `pointer_size - 1` bytes, as that one would still overlap with
+        // the beginning of this range.
+        let start = offset.bytes().saturating_sub(hdl.pointer_size().bytes() - 1);
+        let end = offset + size; // this does overflow checking
+        Ok(self.relocations.range(Size::from_bytes(start)..end))
+    }
+
+    pub fn get_bytes(
+        &self,
+        hdl: impl HasDataLayout,
+        offset: Size,
+        size: Size,
+        required_align: Align,
+    ) -> EvalResult<'tcx, &[u8]> {
+        self.check_align(offset, required_align)?;
+        self.check_bounds(offset, size, true)?;
+        self.check_defined(offset, size)?;
+        self.check_relocations(hdl, offset, size)?;
+        Ok(self.bytes_ignoring_relocations_and_undef(offset, size))
+    }
+
+    pub fn read_bits(
+        &self,
+        tcx: TyCtxt<'_, '_, 'tcx>,
+        offset: Size,
+        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
+    ) -> EvalResult<'tcx, u128> {
+        let ty = tcx.lift_to_global(&ty).unwrap();
+        let layout = tcx.layout_of(ty).unwrap_or_else(|e| {
+            panic!("could not compute layout for {:?}: {:?}", ty, e)
+        });
+        let bytes = self.get_bytes(tcx, offset, layout.size, layout.align)?;
+        Ok(read_target_uint(tcx.data_layout.endian, bytes).unwrap())
+    }
+
+    pub fn read_scalar(
+        &self,
+        hdl: impl HasDataLayout,
+        offset: Size,
+    ) -> EvalResult<'tcx, Scalar<Tag>> {
+        let size = hdl.data_layout().pointer_size;
+        let required_align = hdl.data_layout().pointer_align;
+        self.check_align(offset, required_align)?;
+        self.check_bounds(offset, size, true)?;
+        self.check_defined(offset, size)?;
+        let bytes = self.bytes_ignoring_relocations_and_undef(offset, size);
+        let offset = read_target_uint(hdl.data_layout().endian, &bytes).unwrap();
+        let offset = Size::from_bytes(offset as u64);
+        if let Some(&(tag, alloc_id)) = self.relocations.get(&offset) {
+            Ok(Pointer::new_with_tag(alloc_id, offset, tag).into())
+        } else {
+            Ok(Scalar::Bits {
+                bits: offset.bytes() as u128,
+                size: size.bytes() as u8,
+            })
+        }
+    }
+
+    fn bytes_ignoring_relocations_and_undef(&self, offset: Size, size: Size) -> &[u8] {
```

Comment:
And also ignoring alignment and bounds... wtf? We didn't have anything nearly as unsafe before.